### PR TITLE
Expose service intents

### PR DESCRIPTION
**Pull request by scothis**
An intent is a capability that the application has that needs to be
fulfilled by the platform in order to take effect. For example, an
application that has a mysql driver probably expects to have credentials
for a mysql database bound.

This PR only exposes the intent, it does nothing to either provision an
appropriate service, or bind a provisioned service to the application.
Other tools or aspects of the platform will need to provide those
capabilities.

This change also flips Opinion from a struct to an interface to allow
for different types of opinions. The previous Opinion struct is now a
BasicOpinion. The new SpringBootServiceIntent is an Opinion that looks
for drivers for the service and exposes the intent as a label in a
consistent way.

### CODE REVIEW

File: opinions/opinions.go [hunk 1]
```diff
@@ -54,3 +60,24 @@ func (os AppliedOpinions) Has(id string) bool {
 	}
 	return false
 }
+
+type BasicOpinion struct {
+	Id             string
+	ApplicableFunc func(applied AppliedOpinions, metadata cnb.BuildMetadata) bool
+	ApplyFunc      func(ctx context.Context, podSpec *corev1.PodTemplateSpec, metadata cnb.BuildMetadata) error
```
Comment (ekcasey) [inline opinions/opinions.go:67]: What not just `Applicable`? `ApplicableFunc` feels redundant with the type

File: opinions/spring_boot_opinions.go [hunk 2]
```diff
@@ -285,3 +324,35 @@ func SpringApplicationProperties(ctx context.Context) map[string]string {
...
+		}
+	}
+	return false
+}
+
+func (o *SpringBootServiceIntent) Apply(ctx context.Context, podSpec *corev1.PodTemplateSpec, metadata cnb.BuildMetadata) error {
+	bootMetadata := NewSpringBootBOMMetadata(metadata)
+	for _, d := range bootMetadata.Dependencies {
+		if o.Dependencies.Has(d.Name) {
+			podSpec.Labels[o.LabelName] = d.Name
```
Comment (ekcasey) [inline opinions/spring_boot_opinions.go:352]: When we add support for multiple contains we should add something like a complementary annotation so we know which container wants the service.

**Review by ekcasey — APPROVED** [review]

File: opinions/opinions.go (hunk 1 above)
Comment (scothis) [reply opinions/opinions.go:67]: `Applicable` as a field collides with the interface

Comment (ekcasey) [reply opinions/opinions.go:67]: good point

File: opinions/spring_boot_opinions.go (hunk 2 above)
Comment (scothis) [reply opinions/spring_boot_opinions.go:352]: opened #31 to track

**Pull request merged**